nabs/optimize.py

```diff
@@ -0,0 +1,289 @@
+        Option to average the signal we are reading to limit the affect of
+        noise on our measurements
+
+    Returns
+    -------
+    bounds: tuple
+        The range in which we have determined the extrema to lie within.
+    """
+    # This is boiler plate code and should be packaged into a
+    # pre-processor, for now we repeat it as to not subscribe numerous
+    # streams
+    average = average or 1
+    stream = AverageStream(num=average)
+    yield from bps.subscribe('all', stream)
+    stream.start({'uid': None})
+
+    # Measurement plan
+    def measure_probe(position):
+        # Move motor
+        yield from bps.mv(motor, position)
+        # Return measurement
+        ret = yield from measure_average([signal, motor], average,
+                                         stream=stream)
+        logger.debug("Found a values of %r at %r",
+                     ret[signal.name], position)
+        return ret[signal.name]
+
+    # If we have already found what we are looking for stop the scan
+    (a, b) = limits
+    region_size = b - a
+    if region_size <= tolerance:
+        return (a, b)
+    # Determine the number of steps to converge
+    n = math.ceil(math.log(tolerance/region_size)
+                  / math.log(1/golden_ratio))
+    logger.debug("Beginning golden-section search, "
+                 "narrowing extrema location to %r "
+                 "will require %r steps",
+                 tolerance, n)
+    # Place holders for probe values
+    c = b - region_size/golden_ratio
+    d = a + region_size/golden_ratio
+    # Examine our new probe locations
+    low_probe = yield from measure_probe(c)
+    high_probe = yield from measure_probe(d)
+    # Begin iteratively narrowing range
+    for step in range(n - 1):
+        logger.debug("Iteration %s: Extrema is between %s and %s",
+                     step + 1, a, b)
+        if low_probe < high_probe:
+            # Readjust region of interest
+            b = d
+            d = c
+            high_probe = low_probe
+            region_size /= golden_ratio
+            # Calculate next probe
+            c = b - region_size/golden_ratio
+            # Measure next probe
+            low_probe = yield from measure_probe(c)
+        else:
+            # Readjust region of interest
+            a = c
+            c = d
+            low_probe = high_probe
+            region_size /= golden_ratio
+            # Calculate next probe
+            d = a + region_size/golden_ratio
+            # Measure next probe
+            high_probe = yield from measure_probe(d)
+    # Return the final banding region
+    if low_probe < high_probe:
+        final_region = (a, d)
+    else:
+        final_region = (c, b)
+    logger.debug("Extrema determined to be within %r",
+                 final_region)
+    return final_region
+
+
+# Add the optimization docstring to both minimize and maximize
+maximize.__doc__ += optimize.__doc__
+minimize.__doc__ += optimize.__doc__
```

Comment:
I'm not sure if you want to dive into the weeds, but there's also a way to edit the `__signature__` attribute so it doesn't just say `*args` and `**kwargs` when you use `?` in `ipython`. Not high prio but I thought you should know.

Reply:
Might be nice. I'll look into this when I fill out the docs